### ISSUE TITLE
[UC] Context aware access policies

### ISSUE THREAD

**Issue by elf-pavlik** [w3c/lws-ucs#17]
**As an average user,**
**I want to give apps permissions that take context into account.**
**So that apps I use with some data in one context can't access the same type of data in another context**

## Preconditions:

*What conditions must be in place or assumed before this use case can begin?*

The user is using various apps in their day-to-day interactions. They also interact in different contexts.

## Trigger:

*What (user or system) event or action initiates this use case?*

The user starts using a new app and wants to give it the minimum permissions needed to accomplish their goals.

## Actors:

*Describe the primary actor and any other relevant actors involved in this use case*

* Alice - the user
* Dr. Jones - one of the doctors treating Alice
* Bob - a random person offering hospitality in some themed network

## Distinction:

*What unique challenges or distinguishing factors (like technical issues, user experience needs, workflow integration, etc.) are associated with this use case?*

A user uses multiple applications to work with the same type of data, but they want to restrict access to each app based on context.

## Scenario:

*Describe an ideal or happy-case scenario where this use case would play out as intended.*

Alice follows treatment with Dr. Jones, and exchanges chat messages with them about that treatment.
She uses the SolidMed app to manage her health care, including that treatment.

Alice also rides bicycles as a hobby and uses a hospitality exchange network with other bikers.
On her planned route, she found Bob offering hospitality, and they discussed the details of her stay at Bob's place.
Alice uses the SolidHospEx app to manage her stay with people in themed networks linked to her hobbies.

Alice gives precise permissions to those two apps:
* SolidMed can only access chat messages related to health treatments
* SolidHospEx can only access chat messages related to hospitality exchanges

## Alternative case(s):

*What alternative flows or variations should the system handle for this use case?*

Alice uses specialized SoildMed and SolidHospEx and a third generic chat application - SolidChat. She gives SolidChat context-independent permissions to access all chat messages, which means it can access messages that each of the other two apps can access and any other chat message that Alice has access to.

## Error scenario:

*What unexpected issues or errors might arise, and how should the system handle them?*

* SolidMed can access chat messages about hospitality exchanges or something else unrelated to health treatments
* SolidHospEx can access chat messages about health treatments or something else unrelated to hospitality exchanges

## Acceptance Criteria:

*What conditions or criteria must be met for this use case to be considered successfully handled? What limitations are acceptable?*

* Setting context-aware policies must be intuitive for the user. The filesystem structure may not meet that requirement since many (already most?) people mainly use mobile devices for their web activities.
* Contexts can be defined based on the common relations in specific business/social domains

## References:

*List any relevant resources or examples that could inform this use case, possibly from other domains or solutions.*

Nowadays, in a silo-based web, hospitality services have internal messaging services, similar medical appointment services, and other kinds of services. This ensures that messages from one context are not available in another context. By changing to a user-centric web, similar boundaries need to be preserved, even though all the data can be colocated in the user-owned storage.

A good example of hospitality exchange effort would be #15 @mrkvon

**Comment by hzbarcea** [comment]
@elf-pavlik , you used the template as intended. Thank you! Feel free to add more use cases.

I like the fact that you used the [Alice/Bob metaphor](https://en.wikipedia.org/wiki/Alice_and_Bob). I wish it would be used consistently, but I'll to my best to translate. I was wondering if any of the personas would represent the intent better, eg. a Charlie of Faythe.

Thanks again, keep them coming!

**Comment by mrkvon** [comment]
Allow me to add that I think Charlie should still be able to access the chat with Alice using another app - SolidPhysician - regardless of whether Alice explicitly allowed that app to read data she owns, as long as she granted read access to Charlie. If Charlie switches to another app - SolidDoc - later on, the chat should still work without patients' intervention.

The same applies to Alice switching from SolidHospEx to SolidHosts or whatnot.

**Comment by acoburn** [comment]
In some cases, a user should be able to use arbitrary apps to interact with data. In other cases, especially in regulated industries (e.g. health care, financial data), the allowable apps will need to be carefully controlled, either by the data subject or by the operator of the environment.

**Comment by elf-pavlik** [comment]
I agree with @acoburn, but I think those should be separate use cases:

* The End User controls which application they use to access the data shared with them.
* The Resource Owner controls which application everyone has to use to access the data they share with others.

This UC not only focuses on the former but also emphasizes the need to set access policies based on context, not only the type of data. Similar functionality might be needed in the latter, but I'll leave those use cases for those who have expertise in those environments.

**Comment by elf-pavlik** [comment]
In the first video, from the [mini-series I prepared for TPAC](https://www.w3.org/2024/09/TPAC/group-updates.html#solid) I try to explain that different contexts with different requirements will need to be supported. I think it is possible to support both approaches by having everything explicit and discoverable.
Once again, it would be very helpful if someone having more experience in those regulated industries created an alternative use case to illustrate scenario that would allow capturing relevant requirements.

**Comment by uvdsl** [comment]
This is an interesting discussion. I think this use case is related to UC #29. However, based on @elf-pavlik's comment I think the current UC and UC #29 are distinct.  I want to check with you all if my understanding is correct.

> I agree with @acoburn, but I think those should be separate use cases:
> 
>     * The End User controls which application they use to access the data shared with them.
> 
>     * The Resource Owner controls which application everyone has to use to access the data they share with others.
> 
> 
> This UC not only focuses on the former but also emphasizes the need to set access policies based on context, not only the type of data. Similar functionality might be needed in the latter, but I'll leave those use cases for those who have expertise in those environments.

From this, I gather that this current UC focuses on the former, while UC #29 focuses on the latter. Would you agree on this distinction? Did I miss something?

**Comment by elf-pavlik** [comment]
@uvdsl thank you for linking to #29. Indeed, it looks like you have already created the complementary use case!
#29 might be more general, while this one is a little more specific, at the same time, depending on the existence of a more specific equivalent. I wonder if, in your scenario, you also have more particular cases that require more sophisticated expressivity from access policies. Here, it is a combination of the type of data and its context, which often would be based on (semantic) relation inside of the data itself.

**Comment by uvdsl** [comment]
Thank you, @elf-pavlik, for confirming my understanding! 

Yes, there are particular use cases for restricting users/IDPs/apps respectively. I shall add them to #29 to illustrate.
For example in a tax advisor ecosystem, a company might restrict the IDPs that users may use to authenticate to a set of recognized/vetted IDPs for certified tax advisors. Similarly for the applications. Restricting user access is the usual cloud resource sharing. 

Now, I can see that @mrkvon  might argue that restricting the applications should not be possible to fight vendor lock-in, but this would pose a security risk: If a user chooses a malicious app  to do their taxes (access their private data), then the malicious app can misuse that private information. 

I do not desire vendor lock-in either, and therefore I think that one should _be able to not_ put any restrictions on data access if they want. At the same time, I think we need _to be able to_ restrict access such that we can address the security requirements from certain (heavily regulated) sectors. 

I would also like to point out that it is not at all trivial and clear whose decision it is to restrict access to data. In some instances, the user is to decide while in other instances an ecosystem supervising body (or someone/thing else) is in charge of this decision.

For this use case, I am wondering about the term "context". How would you define context? What elements is the context comprised of? Is it a finite list? Do you imagine a requirement for doing reasoning over the data in question?

**Comment by csarven** [comment]
General note about a recurring pattern in some of the use cases: we should do better than "user". "User" is overly vague for use cases or at the very least obvious because the other possible thing is software (or well hardware, cyborgs etc..). So, we definitely need to describe the actors better:

**actor diversity with variations in needs, capabilities, and aspirations.**

--

I don't think the term "average user" is particularly useful. We should refer to more concrete roles or describe their experiences and knowledge (as in the Actors section). Just because some operating systems on some mobile devices provide a similar "context" for managing access to data or hardware, it does not mean that the "average user" understands anything about it. For instance, I have observed people enabling GPS, microphone access, or even granting access to their entire storage simply because it is far more convenient than being prompted every single time. Another example: the administrators of an organisation often lock down certain portions of the operating system, maintain configurations, or reset logs in specific apps because the institution cannot trust the "average user" to consistently to do the right thing (TM). Unfortunately, we don't live in that world, because "computer" literacy is something we are trying to improve across the board. Lastly, assuming that the "average user" will be able to make correct decisions undermines not only a great portion of actually average users on the web but also those that are not yet on the web! And when I get on the web, this will be a major accessibility barrier on day one.

What you could say is perhaps "As an admin" which would certainly fit better. This is because we can reasonably expect much more from an "admin" than an "average user" who is unlikely to fully grasp the implications of what comes out of this use case.
